### PR TITLE
autoscaling: Implement TidbInformer

### PR DESCRIPTION
**Pull request by Yisaer**
### What problem does this PR solve?
Implement TidbInformer

### Check List

Tests

- Unit test

Related changes

- Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by Yisaer** [comment]
@howardlau1999 PTAL

File: pkg/autoscaling/calculation.go [hunk 1]
```diff
@@ -41,6 +45,12 @@ var (
 	MaxScaleInStep uint64 = 1
 )
 
+func setInformer(rc *cluster.RaftCluster) {
+	if informer == nil {
+		informer = newTidbInformer(rc.GetEtcdClient())
+	}
```
Comment (howardlau1999) [inline pkg/autoscaling/calculation.go:51]: when multiple requests arrive at the same time, is it acceptable to create multiple informer instances here?

**Comment by howardlau1999** [comment]
The rest LGTM

File: pkg/autoscaling/calculation.go (hunk 1 above)
Comment (Yisaer) [reply pkg/autoscaling/calculation.go]: good catch, updated.

File: pkg/autoscaling/calculation.go [hunk 2]
```diff
@@ -288,7 +288,11 @@ func getScaledTiKVGroups(informer core.StoreSetInformer, healthyInstances []inst
 func getScaledTiDBGroups(informer tidbInformer, healthyInstances []instance) []*Plan {
 	planMap := make(map[string]map[string]struct{}, len(healthyInstances))
 	for _, instance := range healthyInstances {
-		tidb := informer.GetTiDB(instance.address)
+		tidb, err := informer.GetTiDB(instance.address)
+		if err != nil {
+			// TODO: error handling
```
Comment (rleungx) [inline pkg/autoscaling/calculation.go:293]: What should we do to handle this kind of error?

File: pkg/autoscaling/types.go [hunk 3]
```diff
@@ -138,3 +148,38 @@ func (t *TiDBInfo) getLabelValue(key string) string {
...
+}
+
+func (informer *tidbInformerImpl) GetTiDB(address string) (*TiDBInfo, error) {
+	key := fmt.Sprintf("/topology/tidb/%s/info", address)
+	resp, err := etcdutil.EtcdKVGet(informer.etcdClient, key)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Count < 1 {
+		err := fmt.Errorf("resp loaded for tidb[%s] is empty", address)
```
Comment (rleungx) [inline pkg/autoscaling/types.go:169]: There is no need to define a new error. We can directly output what we wanna show in the log.

File: pkg/autoscaling/calculation.go (hunk 2 above)
Comment (Yisaer) [reply pkg/autoscaling/calculation.go:293]: I think we could package the error message and tell the caller(deploy layer) the error reason in response.

File: pkg/autoscaling/types.go (hunk 3 above)
Comment (Yisaer) [reply pkg/autoscaling/types.go:181]: updated.

File: pkg/autoscaling/calculation.go [hunk 4]
```diff
@@ -117,9 +125,21 @@ func filterTiKVInstances(informer core.StoreSetInformer) []instance {
 	return instances
 }
 
-// TODO: get TiDB instances, we can directly visit prometheus 'up{job="tidb"}' metrics to know the healthy tidb instances
-func getTiDBInstances() []instance {
-	return []instance{}
+func getTiDBInstances(rc *cluster.RaftCluster) []instance {
+	informerInit.Do(func() {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/autoscaling/calculation.go]: removed.

File: pkg/autoscaling/types.go [hunk 5]
```diff
@@ -134,7 +144,66 @@ func (t *TiDBInfo) getLabelValue(key string) string {
...
+func (informer *tidbInformerImpl) GetTiDBs() ([]*TiDBInfo, error) {
+	prefix := "/topology/tidb/"
+	resps, err := etcdutil.EtcdKVGet(informer.etcdClient, prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+	tidbs := make([]*TiDBInfo, 0, resps.Count)
+	for _, resp := range resps.Kvs {
+		key := string(resp.Key)
+		if tidbInfoPattern.MatchString(key) {
```
[the remark this answers is not in the document]
Comment (Yisaer) [reply pkg/autoscaling/types.go:202]: Using `regexp.MatchString` would be needed to handle error in every loop. I tend to init `tidbInfoPattern` at first.

File: pkg/autoscaling/types.go [hunk 6]
```diff
@@ -138,3 +140,46 @@ func (t *TiDBInfo) getLabelValue(key string) string {
...
+
+// GetTiDB get TiDB info which registered in PD by address
+func GetTiDB(etcdClient *clientv3.Client, address string) (*TiDBInfo, error) {
+	key := fmt.Sprintf("/topology/tidb/%s/info", address)
+	resp, err := etcdutil.EtcdKVGet(etcdClient, key)
+	if err != nil {
+		return nil, err
+	}
+	if resp.Count < 1 {
+		err := fmt.Errorf("resp loaded for tidb[%s] is empty", address)
```
Comment (rleungx) [inline pkg/autoscaling/types.go:152]:
```suggestion
		err := fmt.Errorf("resp loaded for tidb [%s] is empty", address)
```

File: pkg/autoscaling/types.go [hunk 7]
```diff
@@ -138,3 +140,46 @@ func (t *TiDBInfo) getLabelValue(key string) string {
...
+	if err != nil {
+		return nil, err
+	}
+	tidb.Address = address
+	return tidb, nil
+}
+
+// GetTiDBs list TiDB register in PD
+func GetTiDBs(etcdClient *clientv3.Client) ([]*TiDBInfo, error) {
+	tidbInfoPattern, err := regexp.Compile("/topology/tidb/.+/info")
```
Comment (rleungx) [inline pkg/autoscaling/types.go:166]: prefer to use a const for "/topology/tidb/.+/info"

Comment (Yisaer) [reply pkg/autoscaling/types.go]: updated.

**Review by rleungx — APPROVED** [review]
LGTM

File: pkg/autoscaling/types.go [hunk 8]
```diff
@@ -138,3 +140,54 @@ func (t *TiDBInfo) getLabelValue(key string) string {
...
+	prefix := "/topology/tidb/"
+	resps, err := etcdutil.EtcdKVGet(etcdClient, prefix, clientv3.WithPrefix())
+	if err != nil {
+		return nil, err
+	}
+	tidbs := make([]*TiDBInfo, 0, resps.Count)
+	for _, resp := range resps.Kvs {
+		key := string(resp.Key)
+		if tidbInfoPattern.MatchString(key) {
+			address := key[len("/topology/tidb/") : len(key)-5]
```
Comment (howardlau1999) [inline pkg/autoscaling/types.go:183]:
```suggestion
			address := key[len(prefix) : len(key)-len("/info")]
```

**Comment by howardlau1999** [comment]
The rest LGTM

File: pkg/autoscaling/types.go [hunk 9]
```diff
@@ -138,3 +140,54 @@ func (t *TiDBInfo) getLabelValue(key string) string {
...
+	tidbInfoPatternStr = "/topology/tidb/.+/info"
+)
+
+// GetTiDBs list TiDB register in PD
+func GetTiDBs(etcdClient *clientv3.Client) ([]*TiDBInfo, error) {
+	tidbInfoPattern, err := regexp.Compile(tidbInfoPatternStr)
+	if err != nil {
+		return nil, err
+	}
+	prefix := "/topology/tidb/"
```
Comment (howardlau1999) [inline pkg/autoscaling/types.go:174]: How about making it a constant

File: pkg/autoscaling/types.go (hunk 8 above)
Comment (howardlau1999) [reply pkg/autoscaling/types.go:183]: Or you can add a capture group in the regex to directly extract the address

File: pkg/autoscaling/types.go (hunk 9 above)
Comment (Yisaer) [reply pkg/autoscaling/types.go]: updated.

**Comment by howardlau1999** [comment]
LGTM

**Pull request merged**